src/macaron/repo_finder/commit_finder.py

```diff
@@ -0,0 +1,397 @@
+            continue
+
+        valid_tags.append(tag)
+
+    if not valid_tags:
+        logger.debug("No tags with commits found for %s", str(purl))
+        return "", ""
+
+    # Match tags.
+    matched_tags = _match_tags(valid_tags, target_version_pattern, parts, has_non_numeric_suffix, purl.name)
+
+    if not matched_tags:
+        logger.debug("No tags matched for %s", str(purl))
+        return "", ""
+
+    if len(matched_tags) > 1:
+        logger.debug("Tags found for %s: %s", str(purl), len(matched_tags))
+        logger.debug("Best match: %s", matched_tags[0])
+        logger.debug("Up to 5 others: %s", matched_tags[1:6])
+
+    tag = matched_tags[0]
+    tag_name = str(tag)
+    branches = git_obj.get_commit_from_tag(tag_name).branches
+    logger.debug("Branches: %s", branches)
+
+    if not branches:
+        logger.debug("No branch associated with tag (commit): %s (%s)", tag_name, tag.commit.hexsha)
+        return "", ""
+
+    branch_name = ""
+    for branch in branches:
+        # Ensure the detached head branch is not picked up.
+        if "(HEAD detached at" not in branch:
+            branch_name = branch
+            break
+
+    if not branch_name:
+        logger.debug("No valid branch associated with tag (commit): %s (%s)", tag_name, tag.commit.hexsha)
+        return "", ""
+
+    logger.debug(
+        "Found tag %s with commit %s of branch %s for artifact version %s@%s",
+        tag,
+        tag.commit.hexsha,
+        branch_name,
+        purl.name,
+        purl.version,
+    )
+    return branch_name, tag.commit.hexsha
+
+
+def _build_version_pattern(version: str) -> tuple[Pattern, list[str], bool]:
+    """Build a version pattern to match the passed version string.
+
+    Parameters
+    ----------
+    version: str
+        The version string.
+
+    Returns
+    -------
+    tuple[Pattern, list[str], bool]
+        The tuple of the regex pattern that will match the version, the list of version parts that were extracted, and
+        whether the version string has a non-numeric suffix.
+
+    """
+    # The version is split on non-alphanumeric characters to separate the version parts from the non-version parts.
+    # e.g. 1.2.3-DEV -> [1, 2, 3, DEV]
+    split = split_pattern.split(version)
+    logger.debug("Split version: %s", split)
+    if not split:
+        # If the version string contains no separators use it as is.
+        split = [version]
+
+    this_version_pattern = ""
+    parts = []
+    numeric_parts = 0
+    has_non_numeric_suffix = False
+    for count, part in enumerate(split):
+        # Validate the split part by checking it is only comprised of alphanumeric characters.
+        valid = validation_pattern.match(part)
+        if not valid:
+            continue
+        parts.append(part)
+
+        # Detect versions that end with a zero, so the zero can be made optional.
+        has_trailing_zero = len(split) > 2 and count == len(split) - 1 and part == "0"
```

Comment:
I think `len(split)` can be computed outside of the for loop.

Reply:
Do you mean `has_trailing_zero` can be computed outside of the for loop instead?

Reply:
I've moved `has_trailing_zero` out of the loop and simplified it.
`len(split)` is O(1) so there's no real benefit there.